File: packages/vscode-extension/src/chat/chat-events.js
/**
 * Map the CLI's stream-json NDJSON events onto the small message vocabulary
 * the chat webview renders. Pure + stateful-per-turn (the reducer tracks
 * whether text deltas streamed this turn so the final `result` text is not
 * rendered twice). No `vscode`, fully unit-testable.
 *
 * UI message kinds:
 *   init       { model, provider, sessionId }
 *   delta      { text }                    streaming assistant text
 *   thinking   { text }                    streaming extended-thinking reasoning
 *   tool       { tool, summary }           a tool call started
 *   tool_done  { tool, isError }           …finished
 *   info       { text }                    compaction / misc one-liners
 *   turn_end   { isError, text|null, usage } text only when nothing streamed
 *   error      { text }                    session-level failure
 */

/** One-line argument summary for the tool trace (mirrors the CLI's trace). */
function summarizeToolArgs(args) {
  if (!args || typeof args !== "object") return "";
  const s =
    args.path ||
    args.command ||
    args.pattern ||
    args.query ||
    args.url ||
    args.code ||
    "";
  const str = String(s);
  return str.length > 80 ? str.slice(0, 80) + "…" : str;
}

function createTurnState() {
  return { sawDelta: false };
}

/**
 * @param {object} evt    one parsed NDJSON event from the agent child
 * @param {object} state  per-conversation state from createTurnState()
 * @returns {object|null} a UI message, or null when the event is UI-silent
 */
function mapAgentEvent(evt, state) {
  if (!evt || typeof evt !== "object") return null;
  switch (evt.type) {
    case "system":
      if (evt.subtype === "init") {
        return {
          kind: "init",
          model: evt.model || "",
          provider: evt.provider || "",
          sessionId: evt.session_id || "",
        };
      }
      return null;
    case "stream_event": {
      const delta = evt.event?.delta;
      if (delta?.type === "text_delta" && typeof delta.text === "string") {
        state.sawDelta = true;
        return { kind: "delta", text: delta.text };
      }
      // Extended-thinking reasoning (Anthropic, when /think is on) — rendered
      // as a dimmed/collapsed block, separate from the answer text.
      if (
        delta?.type === "thinking_delta" &&
        typeof delta.thinking === "string"
      ) {
        return { kind: "thinking", text: delta.thinking };
      }
      return null;
    }
    case "tool_use":
      return {
        kind: "tool",
        tool: evt.tool || "?",
        summary: summarizeToolArgs(evt.args),
      };
    case "tool_result": {
      // `ask_user_question` normally round-trips through the in-panel question
      // card (question_request below; the panel always opts in via
      // CC_INTERACTIVE_QUESTIONS=1). An old `cc` that ignores the env var — or a
      // question the user never answers — degrades gracefully instead: the
      // handler returns {error:"user_not_reachable"} (or "user_timeout") and the
      // model proceeds autonomously. That is NOT a tool failure — don't paint a
      // scary red "✗ … failed"; surface a quiet note.
      const errText =
        typeof evt.error === "string"
          ? evt.error
          : evt.result && typeof evt.result.error === "string"
            ? evt.result.error
            : null;
      const benign =
        errText === "user_not_reachable" || errText === "user_timeout";
      return {
        kind: "tool_done",
        tool: evt.tool || "?",
        isError: evt.is_error === true && !benign,
        note: benign
          ? evt.tool === "ask_user_question"
            ? "couldn't ask interactively in the panel — proceeding autonomously"
            : "skipped — proceeding"
          : null,
      };
    }
    case "compaction":
      return {
        kind: "info",
        text: `compacted: saved ${evt.stats?.saved ?? "?"} tokens`,
      };
    case "result": {
      const sawDelta = state.sawDelta;
      state.sawDelta = false; // reset for the next turn
      if (evt.subtype === "interrupted" || evt.interrupted === true) {
        return {
          kind: "turn_end",
          isError: false,
          text: "⏹ interrupted",
          usage: null,
        };
      }
      // Budget stops (error_max_turns / error_max_budget) carry the FULL
      // final text in evt.result — repainting it as the error would duplicate
      // everything that already streamed, in red. Show the stop reason
      // instead (the dedicated budget events above carry the figures).
      if (
        evt.subtype === "error_max_turns" ||
        evt.subtype === "error_max_budget"
      ) {
        const reason =
          evt.subtype === "error_max_turns"
            ? "⏹ stopped: turn budget exhausted"
            : "⏹ stopped: cost budget exhausted";
        return {
          kind: "turn_end",
          isError: true,
          text: sawDelta
            ? reason
            : evt.result
              ? `${evt.result}\n${reason}`
              : reason,
          usage: evt.usage || null,
        };
      }
      return {
        kind: "turn_end",
        isError: evt.is_error === true,
        // If text streamed via deltas, the final result repeats it — drop it.
        text: evt.is_error
          ? evt.error || evt.result || "turn failed"
          : sawDelta
            ? null
            : evt.result || null,
        usage: evt.usage || null,
      };
    }
    case "approval_request":
      return {
        kind: "approval",
        id: evt.id,
        tool: evt.tool || null,
        command: evt.command || null,
        risk: evt.risk || null,
        rule: evt.rule || null,
        reason: evt.reason || null,
      };
    case "approval_resolved":
      return {
        kind: "approval_done",
        id: evt.id,
        approved: evt.approved === true,
        via: evt.via || null,
      };
    case "question_request":
      // ask_user_question round-trip (CC_INTERACTIVE_QUESTIONS): the agent is
      // BLOCKED waiting for the user. chat-view shows a native QuickPick and
      // replies {type:"answer",...}; the webview renders the question inline.
      const question = {
        kind: "question",
        id: evt.id,
        question: evt.question || "",
        options: Array.isArray(evt.options) ? evt.options : null,
        multiSelect: evt.multiSelect === true,
      };
      if (evt.metadata && typeof evt.metadata === "object") {
        question.metadata = evt.metadata;
        question.elicitation = evt.metadata.kind === "mcp_elicitation";
        question.requestedSchema =
          evt.metadata.requestedSchema || evt.requestedSchema || null;
        question.server = evt.metadata.server || null;
      }
      return question;
    case "question_resolved":
      return {
        kind: "info",
        text:
          evt.via === "user-answer"
            ? "✓ answered"
            : "ask_user_question: no answer — proceeding",
      };
    case "plan_update":
      return {
        kind: "plan",
        active: evt.active === true,
        state: evt.state || null,
        plan_id: evt.plan_id || null,
        plan_version: evt.plan_version || null,
        previous_plan_id: evt.previous_plan_id || null,
        items: Array.isArray(evt.items) ? evt.items : [],
        risk: evt.risk || null,
        execution_lock: evt.execution_lock || null,
        note: evt.note || null,
      };
    case "session_error":
      return { kind: "error", text: evt.error || "agent session error" };
    case "token_usage":
      // Per-LLM-call usage mid-turn — the webview accumulates these into a live
      // token counter on the status line (turn_end still carries the
      // authoritative turn total from the result envelope).
      return evt.usage && typeof evt.usage === "object"
        ? { kind: "usage", usage: evt.usage }
        : null;
    case "iteration_warning":
      return {
        kind: "info",
        text: "⚠ " + (evt.message || "approaching the iteration limit"),
      };
    case "stream_retry":
      // Transient API connection drop mid-stream — without this line the
      // reconnect loop is an unexplained "thinking…" stall.
      return {
        kind: "info",
        text:
          "⚠ " +
          (evt.message || "API connection dropped — retrying") +
          (Number.isFinite(evt.attempt) ? ` (attempt ${evt.attempt})` : ""),
      };
    case "iteration_budget_exhausted":
      return {
        kind: "info",
        text:
          "⏹ turn budget exhausted" +
          (Number.isFinite(evt.budget) ? ` (${evt.budget} turns)` : ""),
      };
    case "cost_budget_exhausted": {
      const spent = Number.isFinite(evt.spent_usd)
        ? `$${Number(evt.spent_usd).toFixed(2)}`
        : null;
      const limit = Number.isFinite(evt.limit_usd)
        ? `$${Number(evt.limit_usd).toFixed(2)}`
        : null;
      return {
        kind: "info",
        text:
          "⏹ cost budget exhausted" +
          (spent && limit ? ` (${spent} of ${limit})` : ""),
      };
    }
    case "checkpoint":
      // Auto-snapshot taken before a mutating tool (git shadow-commit). A terse
      // trace so `/rewind` has a visible anchor; the tool gives it context.
      return {
        kind: "info",
        text:
          "📸 snapshot" +
          (typeof evt.tool === "string" && evt.tool
            ? ` before ${evt.tool}`
            : ""),
      };
    case "raw":
      return { kind: "info", text: evt.text };
    default:
      return null; // unknown/UI-silent event types
  }
}

/**
 * Permission modes the panel can pass through to `cc agent --permission-mode`
 * (the panel's auto-accept / bypass selector — Claude-Code parity). "default"
 * (and anything unknown) emits NO flag, so the agent keeps its normal approval
 * flow. `plan` is accepted here for completeness, but the panel drives plan
 * mode live over the stdin protocol (the `plan` event), not at spawn time.
 */
const PERMISSION_MODES = new Set(["plan", "acceptEdits", "bypassPermissions"]);

/**
 * Extra CLI args for the chat child from user settings + stored session.
 * Empty/missing values fall through to the CLI's own config defaults
 * (pure; vscode-free). `resume` is the workspace's last chat session id —
 * passing it makes the CLI rebuild that conversation and keep appending to
 * it, so the panel survives restarts/reloads with full context. `mode` is the
 * conversation's approval mode (acceptEdits / bypassPermissions); "default"
 * or unset adds nothing. `think` toggles Anthropic extended thinking:
 * "on" → --think, "ultra" → --ultrathink (ignored by non-Anthropic providers);
 * "off"/unset adds nothing.
 */
/** Read the user's cc LLM config — the `llm` block of ~/.chainlesschain/config.json. */
function readCcLlmConfig() {
  try {
    const fs = require("fs");
    const os = require("os");
    const path = require("path");
    const p = path.join(os.homedir(), ".chainlesschain", "config.json");
    return (JSON.parse(fs.readFileSync(p, "utf8")) || {}).llm || {};
  } catch {
    return {}; // no/unreadable config → let the CLI resolve on its own
  }
}

/**
 * Resolve the panel's EFFECTIVE provider/model. A non-empty panel override
 * (`chainlesschain.chat.provider`/`.model`) always wins; when empty, fall back
 * to the user's cc config (~/.chainlesschain/config.json) and return it so the
 * caller can pass it EXPLICITLY as `--provider/--model`.
 *
 * Why explicit instead of letting the empty case fall through to the CLI's own
 * config resolution: the panel spawns `cc` with the IDE's full environment and
 * working directory, and a stale/ambient state there could otherwise make the
 * child resolve a different provider than the terminal — surfacing as a
 * spurious "Anthropic error: 401" when the user actually configured, say,
 * volcengine. Pinning the config value makes the panel deterministically match
 * `cc config`. `readConfig` is injectable for tests.
 */
function resolveChatLlm(
  { provider, model } = {},
  readConfig = readCcLlmConfig,
) {
  const llm = readConfig() || {};
  let p = typeof provider === "string" ? provider.trim() : "";
  let m = typeof model === "string" ? model.trim() : "";
  let baseUrl = "";
  let apiKey = "";
  if (!p && llm.provider) {
    p = String(llm.provider);
    if (!m && llm.model) m = String(llm.model);
  }
  // Carry the FULL llm block (baseUrl + apiKey, + model) when the effective
  // provider matches the configured one. The panel pins --provider/--model and
  // MUST also pass the endpoint + key: the CLI, seeing an explicit --provider,
  // skips config resolution and would otherwise drop a cloud provider's
  // baseUrl/key → the endpoint falls through to ollama ("配置了火山却 fetch
  // failed / 切到 ollama"). Same provider → its baseUrl/apiKey are exactly
  // right. A DIFFERENT explicit provider override carries neither (the user
  // owns that endpoint/key).
  if (p && llm.provider && p === String(llm.provider)) {
    if (!m && llm.model) m = String(llm.model);
    if (llm.baseUrl) baseUrl = String(llm.baseUrl);
    if (llm.apiKey) apiKey = String(llm.apiKey);
  }
  return { provider: p, model: m, baseUrl, apiKey };
}

function buildSessionArgs({
  model,
  provider,
  baseUrl,
  apiKey,
  resume,
  mode,
  think,
  goalCondition,
} = {}) {
  const args = [];
  if (typeof provider === "string" && provider.trim()) {
    args.push("--provider", provider.trim());
  }
  if (typeof model === "string" && model.trim()) {
    args.push("--model", model.trim());
  }
  // Pass the endpoint + key alongside the provider so the CLI doesn't drop a
  // cloud provider's credentials (it skips config when --provider is explicit).
  if (typeof baseUrl === "string" && baseUrl.trim()) {
    args.push("--base-url", baseUrl.trim());
  }
  if (typeof apiKey === "string" && apiKey.trim()) {
    args.push("--api-key", apiKey.trim());
  }
  if (typeof resume === "string" && resume.trim()) {
    args.push("--resume", resume.trim());
  }
  if (typeof mode === "string" && PERMISSION_MODES.has(mode.trim())) {
    args.push("--permission-mode", mode.trim());
  }
  const t = typeof think === "string" ? think.trim() : "";
  if (t === "ultra") args.push("--ultrathink");
  else if (t === "on") args.push("--think");
  if (typeof goalCondition === "string" && goalCondition.trim()) {
    args.push("--goal-condition", goalCondition.trim());
  }
  return args;
}

function buildIdeToolAdmission(source = "vscode-extension") {
  return {
    enforce: true,
    source,
    capabilityGranted: true,
    policyAllowed: true,
    permissionGranted: true,
    budgetOk: true,
    uiSupported: true,
    tools: {
      publish_artifact: { policyAllowed: false },
      notify: { policyAllowed: false },
    },
  };
}

module.exports = {
  mapAgentEvent,
  createTurnState,
  summarizeToolArgs,
  buildSessionArgs,
  buildIdeToolAdmission,
  resolveChatLlm,
  readCcLlmConfig,
  PERMISSION_MODES,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/ChatEvents.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Map the CLI's stream-json NDJSON events onto the small message vocabulary
 * the chat tool window renders — the Java twin of the VS Code extension's
 * {@code chat/chat-events.js} (same kinds, same delta-dedup rule, so the two
 * IDE panels stay behaviorally identical).
 *
 * UI message kinds: init / delta / tool / tool_done / info / turn_end /
 * approval / approval_done / plan / error. Pure JDK, headless-testable.
 */
public final class ChatEvents {
    private ChatEvents() {}

    /** Per-conversation state: did text stream via deltas this turn? */
    public static final class TurnState {
        public boolean sawDelta;
    }

    /**
     * Live per-turn token tally: {@code token_usage} fires once per LLM call
     * while the agent works; the panel accumulates them into a status line
     * ("thinking… · 12.3k→456 tokens (900 cached)"). The turn's {@code result}
     * envelope still carries the authoritative total ({@link #readyLine}).
     * Pure JDK — mirrors the VS Code webview's turnTokens accumulator.
     */
    public static final class TokenTally {
        public long input;
        public long output;
        public long cached;

        /** Fold one token_usage payload in (missing/non-numeric fields = 0). */
        public void add(Map<String, Object> usage) {
            if (usage == null) return;
            input += num(usage.get("input_tokens"));
            output += num(usage.get("output_tokens"));
            cached += num(usage.get("cache_read_input_tokens"));
        }

        /** The mid-turn status line, e.g. "thinking… · 12.3k→456 tokens (900 cached)". */
        public String statusLine() {
            return "thinking… · " + formatTokens(input) + "→" + formatTokens(output)
                    + " tokens"
                    + (cached > 0 ? " (" + formatTokens(cached) + " cached)" : "");
        }
    }

    /** Compact token count, VS Code tokfmt twin: 456 → "456", 12345 → "12.3k",
     *  123456 → "123k" (≥10k drops the decimal). */
    public static String formatTokens(long n) {
        if (n < 1000) return String.valueOf(n);
        double k = n / 1000.0;
        return (n >= 10_000
                ? String.valueOf(Math.round(k))
                : String.format(java.util.Locale.ROOT, "%.1f", k)) + "k";
    }

    /** The end-of-turn status line from the result envelope's usage, e.g.
     *  "ready · 12.3k→456 tokens" — or plain "ready" without usage. */
    public static String readyLine(Map<String, Object> usage) {
        if (usage == null) return "ready";
        return "ready · " + formatTokens(num(usage.get("input_tokens")))
                + "→" + formatTokens(num(usage.get("output_tokens"))) + " tokens";
    }

    private static long num(Object v) {
        return v instanceof Number ? ((Number) v).longValue() : 0L;
    }

    /** One-line argument summary for the tool trace (mirrors the CLI). */
    public static String summarizeToolArgs(Object args) {
        if (!(args instanceof Map)) return "";
        Map<?, ?> m = (Map<?, ?>) args;
        Object s = first(m, "path", "command", "pattern", "query", "url", "code");
        if (s == null) return "";
        String str = String.valueOf(s);
        return str.length() > 80 ? str.substring(0, 80) + "…" : str;
    }

    private static Object first(Map<?, ?> m, String... keys) {
        for (String k : keys) {
            Object v = m.get(k);
            if (v != null && !"".equals(v)) return v;
        }
        return null;
    }

    private static String str(Map<String, Object> m, String key, String dflt) {
        Object v = m.get(key);
        return v == null ? dflt : String.valueOf(v);
    }

    private static boolean isTrue(Object v) {
        return Boolean.TRUE.equals(v);
    }

    private static Map<String, Object> ui(String kind) {
        Map<String, Object> m = MiniJson.obj();
        m.put("kind", kind);
        return m;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> asMap(Object v) {
        return v instanceof Map ? (Map<String, Object>) v : null;
    }

    /**
     * @return a UI message map (key {@code kind}), or null when the event is
     *         UI-silent (iteration_budget_exhausted, …).
     */
    public static Map<String, Object> mapAgentEvent(
            Map<String, Object> evt, TurnState state) {
        if (evt == null) return null;
        String type = str(evt, "type", "");
        if ("system".equals(type)) {
            if ("init".equals(str(evt, "subtype", ""))) {
                Map<String, Object> m = ui("init");
                m.put("model", str(evt, "model", ""));
                m.put("provider", str(evt, "provider", ""));
                m.put("sessionId", str(evt, "session_id", ""));
                return m;
            }
            return null;
        }
        if ("stream_event".equals(type)) {
            Map<String, Object> event = asMap(evt.get("event"));
            Map<String, Object> delta = event == null ? null : asMap(event.get("delta"));
            if (delta != null
                    && "text_delta".equals(str(delta, "type", ""))
                    && delta.get("text") instanceof String) {
                state.sawDelta = true;
                Map<String, Object> m = ui("delta");
                m.put("text", delta.get("text"));
                return m;
            }
            // Extended-thinking reasoning delta (Anthropic; only when /think is on)
            // → a dimmed "thinking" line, like the VS Code panel.
            if (delta != null
                    && "thinking_delta".equals(str(delta, "type", ""))
                    && delta.get("thinking") instanceof String) {
                Map<String, Object> m = ui("thinking");
                m.put("text", delta.get("thinking"));
                return m;
            }
            return null;
        }
        if ("tool_use".equals(type)) {
            Map<String, Object> m = ui("tool");
            m.put("tool", str(evt, "tool", "?"));
            m.put("summary", summarizeToolArgs(evt.get("args")));
            return m;
        }
        if ("tool_result".equals(type)) {
            Map<String, Object> m = ui("tool_done");
            String tool = str(evt, "tool", "?");
            m.put("tool", tool);
            // ask_user_question degrades gracefully in the panel (no interactive
            // round-trip yet): {error:"user_not_reachable"|"user_timeout"} and the
            // model proceeds autonomously — that is NOT a tool FAILURE, so don't
            // paint a scary "✗ … failed"; surface a quiet note instead.
            Object errO = evt.get("error");
            String errText = errO instanceof String ? (String) errO : null;
            if (errText == null) {
                Map<String, Object> res = asMap(evt.get("result"));
                Object re = res == null ? null : res.get("error");
                if (re instanceof String) errText = (String) re;
            }
            boolean benign = "user_not_reachable".equals(errText)
                    || "user_timeout".equals(errText);
            m.put("isError", isTrue(evt.get("is_error")) && !benign);
            if (benign) {
                m.put("note", "ask_user_question".equals(tool)
                        ? "面板暂不支持交互提问 —— 已按最佳判断继续"
                        : "已跳过 —— 继续");
            }
            return m;
        }
        if ("compaction".equals(type)) {
            Map<String, Object> stats = asMap(evt.get("stats"));
            Object saved = stats == null ? null : stats.get("saved");
            Map<String, Object> m = ui("info");
            m.put("text", "compacted: saved " + (saved == null ? "?" : saved) + " tokens");
            return m;
        }
        if ("result".equals(type)) {
            boolean sawDelta = state.sawDelta;
            state.sawDelta = false; // reset for the next turn
            Map<String, Object> m = ui("turn_end");
            if ("interrupted".equals(str(evt, "subtype", ""))
                    || isTrue(evt.get("interrupted"))) {
                m.put("isError", false);
                m.put("text", "⏹ interrupted");
                m.put("usage", null);
                return m;
            }
            boolean isError = isTrue(evt.get("is_error"));
            m.put("isError", isError);
            Object text;
            if (isError) {
                text = evt.get("error") != null ? evt.get("error")
                        : evt.get("result") != null ? evt.get("result") : "turn failed";
            } else {
                // If text streamed via deltas, the final result repeats it — drop.
                text = sawDelta ? null : evt.get("result");
            }
            m.put("text", text);
            m.put("usage", evt.get("usage"));
            return m;
        }
        if ("approval_request".equals(type)) {
            Map<String, Object> m = ui("approval");
            m.put("id", evt.get("id"));
            m.put("tool", evt.get("tool"));
            m.put("command", evt.get("command"));
            m.put("risk", evt.get("risk"));
            m.put("rule", evt.get("rule"));
            m.put("reason", evt.get("reason"));
            return m;
        }
        if ("approval_resolved".equals(type)) {
            Map<String, Object> m = ui("approval_done");
            m.put("id", evt.get("id"));
            m.put("approved", isTrue(evt.get("approved")));
            m.put("via", evt.get("via"));
            return m;
        }
        if ("question_request".equals(type)) {
            // ask_user_question round-trip (CC_INTERACTIVE_QUESTIONS): the agent is
            // BLOCKED on the user. ConversationView pops a dialog and replies
            // {type:"answer",id,answer}.
            Map<String, Object> m = ui("question");
            m.put("id", evt.get("id"));
            m.put("question", str(evt, "question", ""));
            m.put("options", evt.get("options") instanceof List ? evt.get("options") : null);
            m.put("multiSelect", isTrue(evt.get("multiSelect")));
            return m;
        }
        if ("question_resolved".equals(type)) {
            Map<String, Object> m = ui("info");
            m.put("text", "user-answer".equals(String.valueOf(evt.get("via")))
                    ? "✓ answered"
                    : "ask_user_question: no answer — proceeding");
            return m;
        }
        if ("plan_update".equals(type)) {
            Map<String, Object> m = ui("plan");
            m.put("active", isTrue(evt.get("active")));
            m.put("state", evt.get("state"));
            m.put("items", evt.get("items") instanceof List ? evt.get("items") : new ArrayList<>());
            m.put("risk", evt.get("risk"));
            m.put("note", evt.get("note"));
            return m;
        }
        if ("token_usage".equals(type)) {
            // Per-LLM-call usage mid-turn — the panel accumulates these into a
            // live token counter on the status line (turn_end still carries the
            // authoritative turn total). Same contract as the VS Code panel.
            Map<String, Object> usage = asMap(evt.get("usage"));
            if (usage == null) return null;
            Map<String, Object> m = ui("usage");
            m.put("usage", usage);
            return m;
        }
        if ("iteration_warning".equals(type)) {
            Map<String, Object> m = ui("info");
            Object msg = evt.get("message");
            m.put("text", "⚠ " + (msg instanceof String && !((String) msg).isEmpty()
                    ? msg : "approaching the iteration limit"));
            return m;
        }
        if ("session_error".equals(type)) {
            Map<String, Object> m = ui("error");
            Object err = evt.get("error");
            m.put("text", err == null ? "agent session error" : String.valueOf(err));
            return m;
        }
        if ("raw".equals(type)) {
            Map<String, Object> m = ui("info");
            m.put("text", str(evt, "text", ""));
            return m;
        }
        return null;
    }

    /**
     * Extra CLI args from settings + stored session. Empty values fall through
     * to the CLI's own config defaults; {@code resume} makes the CLI rebuild
     * that conversation so the panel survives IDE restarts with full context.
     */
    public static List<String> buildSessionArgs(
            String provider, String model, String resume) {
        List<String> args = new ArrayList<>();
        if (provider != null && !provider.trim().isEmpty()) {
            args.add("--provider");
            args.add(provider.trim());
        }
        if (model != null && !model.trim().isEmpty()) {
            args.add("--model");
            args.add(model.trim());
        }
        if (resume != null && !resume.trim().isEmpty()) {
            args.add("--resume");
            args.add(resume.trim());
        }
        return args;
    }
}
